src/hotspot/share/gc/parallel/psCardTable.cpp

```diff
@@ -277,7 +292,28 @@ void PSCardTable::scavenge_contents_parallel(ObjectStartArray* start_array,
     {
       // Identify right ends.
       HeapWord* obj_addr = start_array->object_start(cur_stripe_end_addr - 1);
-      HeapWord* obj_end_addr = obj_addr + cast_to_oop(obj_addr)->size();
+      size_t obj_sz = cast_to_oop(obj_addr)->size();
+      HeapWord* obj_end_addr = obj_addr + obj_sz;
+      // Scan the elements of a large array to the stripe end.
+      if (obj_sz >= large_obj_arr_min_words() && cast_to_oop(obj_addr)->is_objArray()) {
+        if (first_obj_addr >= cur_stripe_end_addr) {
+          // Nothing to scan according to constraints given above.
+          // We reach here only for the last stripe below space_top where
+          // object_starts_in_range() returns true if space_top is not aligned to
+          // _card_size and an object starts there.
+          assert(first_obj_addr == space_top, "assumption");
+          assert(cur_stripe_end_addr == space_top, "assumption");
+          // This stripe accomodates the end of a large array. It is scanned by the
+          // thread owning the previous stripe.
+          continue;
+        }
+        // Reaching here we know that the large array starts in this stripe.
+        // If it starts here then its end has to be in a following stripe.
+        assert(obj_addr >= cur_stripe_addr &&
+               obj_end_addr >= cur_stripe_end_addr, "overlapping work");
+        obj_end_addr = cur_stripe_end_addr;
+        large_arr = objArrayOop(cast_to_oop(obj_addr));
```

Comment:
I would almost prefer that the code, on detection of a large objArray that covers the rest of the stripe, just called the "handle-large-object" method in some way.
The remaining code is rather complicated to follow, and if that remaining code had the precondition of not being about handling large objArrays, would be easier to understand.